scripts/viewer.js

```diff
@@ -0,0 +1,150 @@
+
+module.exports = async (callback) => {
+  try {
+    const exchangePromise = getExchange(web3)
+    let bracketAddresses
+    if (argv.masterSafe) {
+      bracketAddresses = await getDeployedBrackets(argv.masterSafe)
+      console.log(`Found ${bracketAddresses.length} brackets`)
+    } else {
+      if (!argv.brackets) {
+        callback("One of --masterSafe or --brackets=... is required.")
+      }
+      bracketAddresses = argv.brackets
+    }
+    const exchange = await exchangePromise
+
+    const detailedBrackets = await Promise.all(
+      bracketAddresses.map(async (bracketAddress) => {
+        const orders = decodeOrders(await exchange.getEncodedUserOrders.call(bracketAddress))
+        let tradedTokenIds = []
+        for (const order of orders) {
+          tradedTokenIds.push(order.buyToken, order.sellToken)
+        }
+        tradedTokenIds = uniqueItems(tradedTokenIds)
+        const tokenInfoPromises = fetchTokenInfoFromExchange(exchange, tradedTokenIds)
+        return {
+          bracketAddress,
+          tokenIds: tradedTokenIds,
+          tokenInfoPromises,
+        }
+      })
+    )
+    console.log("Recovering token balances per bracket...")
+    const tokenBalancesPerUser = await Promise.all(
+      detailedBrackets.map(async ({ bracketAddress, tokenIds, tokenInfoPromises }) => {
+        const tokenBalances = []
+        for (const tokenId of tokenIds) {
+          const tokenData = await tokenInfoPromises[tokenId]
+          const [availableForTrading, requestedForWithdraw, storedInBracket] = await Promise.all([
+            exchange.getBalance(bracketAddress, tokenData.address),
+            getWithdrawableAmount(bracketAddress, tokenData.address, exchange, web3),
+            tokenData.instance.balanceOf(bracketAddress),
+          ])
+          tokenBalances.push({
+            bracketAddress: bracketAddress,
+            tokenId,
+            tokenData,
+            totalBalance: availableForTrading.add(requestedForWithdraw).add(storedInBracket),
+            availableForTrading,
+            requestedForWithdraw,
+            storedInBracket,
+          })
+        }
+        return tokenBalances
+      })
+    )
+
+    const balances = []
+    for (const tokenBalances of tokenBalancesPerUser) {
+      balances.push(...tokenBalances)
+    }
+
+    const totalBalanceSum = {}
+    const requestedForWithdrawSum = {}
+    const storedInBracketSum = {}
+    for (const balanceItem of balances) {
+      if (balanceItem.totalBalance.isZero()) {
+        continue
+      }
+      totalBalanceSum[balanceItem.tokenId] = (totalBalanceSum[balanceItem.tokenId] || new BN(0)).add(balanceItem.totalBalance)
+      requestedForWithdrawSum[balanceItem.tokenId] = (requestedForWithdrawSum[balanceItem.tokenId] || new BN(0)).add(
+        balanceItem.requestedForWithdraw
+      )
+      storedInBracketSum[balanceItem.tokenId] = (storedInBracketSum[balanceItem.tokenId] || new BN(0)).add(
+        balanceItem.storedInBracket
+      )
```

Comment:
Maybe just a nit, but given that JS doesn't have default dict like in python: you may want to consider using the following more common technique for making your accumulators:

https://github.com/gnosis/dex-liquidity-provision/blob/154b7c5a838c049d0d7a316f1b19decc642e9081/scripts/utils/trading_strategy_helpers.js#L449-L457

See definition and implementation of `cummulativeAmounts` as a Map initialized to zeroes. Of course this requires knowing all the unique keys before hand.

Reply:
I tried to make this work but there are two issues with this approach:
1. you can't specify a default for `Map.get`
2. because of 1, you need to know all keys of the map in advance to default them to zero, leading to quite some extra code.

The advantage of `Map` is that you don't risk to overwrite some preexisting prototype keys, but in this case all keys are integers, so there is no risk of a conflict.

So, overall I'd keep the current implementation to avoid the extra code.